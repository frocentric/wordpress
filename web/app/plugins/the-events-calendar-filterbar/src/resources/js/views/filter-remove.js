/* global tribe */
/* eslint-disable no-var, strict */

/**
 * Makes sure we have all the required levels on the Tribe Object.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar = tribe.filterBar || {};

/**
 * Configures Filter Remove Object in the Global Tribe variable.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar.filterRemove = {};

/**
 * Initializes in a Strict env the code that manages the filter remove.
 *
 * @since 5.0.0
 *
 * @param  {PlainObject} $   jQuery
 * @param  {PlainObject} obj tribe.filterBar.filterRemove
 *
 * @return {void}
 */
( function( $, obj ) {
	'use strict';
	var $document = $( document );

	/**
	 * Selectors used for configuration and setup.
	 *
	 * @since 5.0.0
	 *
	 * @type {PlainObject}
	 */
	obj.selectors = {
		removeButton: '[data-js="tribe-filter-bar-c-pill__remove-button"]',
		pillFilterName: '[data-filter-name]',
	};

	/**
	 * Handler for remove button click event.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object of click event.
	 *
	 * @return {void}
	 */
	obj.handleRemoveClick = function( event ) {
		var $removeButton = event.data.target;
		var $container = event.data.container;
		var $pill = $removeButton.closest( '[data-filter-name]' );

		// Return early if pill with data-filter-name attribute is not found.
		if ( ! $pill.length ) {
			return;
		}

		var name = $pill.data( 'filter-name' );

		// Return early if name doesn't exist.
		if ( ! name ) {
			return;
		}

		var location = tribe.filterBar.filters.removeKeyValueFromQuery( window.location, name, true );

		tribe.filterBar.filters.submitRequest( $container, location.href );
	};

	/**
	 * Unbind events for filter remove functionality.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.unbindEvents = function( $container ) {
		$container.find( obj.selectors.removeButton ).off();
	};

	/**
	 * Bind events for filter remove functionality.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.bindEvents = function( $container ) {
		$container
			.find( obj.selectors.removeButton )
			.each( function( index, removeButton ) {
				var $removeButton = $( removeButton );
				$removeButton.on( 'click', { target: $removeButton, container: $container }, obj.handleRemoveClick );
			} );
	};

	/**
	 * Deinitialize filter remove JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object for 'beforeAjaxSuccess.tribeEvents' event.
	 *
	 * @return {void}
	 */
	obj.deinit = function( event ) {
		var $container = event.data.container;
		obj.unbindEvents( $container );
		$container.off( 'beforeAjaxSuccess.tribeEvents', obj.deinit );
	};

	/**
	 * Initialize filter remove JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event}   event      event object for 'afterSetup.tribeEvents' event.
	 * @param  {integer} index      jQuery.each index param from 'afterSetup.tribeEvents' event.
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.init = function( event, index, $container ) {
		obj.bindEvents( $container );
		$container.on( 'beforeAjaxSuccess.tribeEvents', { container: $container }, obj.deinit );
	};

	/**
	 * Handles the initialization of filter remove when Document is ready.
	 *
	 * @since 5.0.0
	 *
	 * @return {void}
	 */
	obj.ready = function() {
		$document.on( 'afterSetup.tribeEvents', tribe.events.views.manager.selectors.container, obj.init );
	};

	// Configure on document ready.
	$( obj.ready );
} )( jQuery, tribe.filterBar.filterRemove );
