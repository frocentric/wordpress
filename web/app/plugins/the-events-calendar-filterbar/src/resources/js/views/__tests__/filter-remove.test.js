describe( 'Filter Remove', () => {
	beforeAll( () => {
		String.prototype.className = function() {
			if (
				(
					'string' !== typeof this &&
					! this instanceof String /* eslint-disable-line no-unsafe-negation */
				) ||
				'function' !== typeof this.replace
			) {
				return this;
			}

			return this.replace( '.', '' );
		};

		global.tribe = {};
		require( '../filter-remove' );
	} );

	afterAll( () => {
		delete String.prototype.className;
		delete global.tribe;
	} );

	describe( 'Selectors', () => {
		test( 'Should match snapshot', () => {
			const selectors = JSON.stringify( tribe.filterBar.filterRemove.selectors );
			expect( selectors ).toMatchSnapshot();
		} );
	} );

	describe( 'Handle remove click', () => {
		let windowHold;

		beforeEach( () => {
			windowHold = global.window;
			delete global.window.location;
			global.window = Object.create( window );
			// url = 'https://test.tri.be/events/month/?range=0-50'
			global.window.location = {
				href: 'https://test.tri.be/events/month/?tribe_eventcategory[0]=hello&tribe_eventcategory[1]=world&tribe_cost=0-100&tribe_custom=moderntribe',
				origin: 'https://test.tri.be',
				pathname: '/events/month/',
				search: '?tribe_eventcategory[0]=hello&tribe_eventcategory[1]=world&tribe_cost=0-100&tribe_custom=moderntribe',
				hash: '',
			};
			global.tribe.filterBar.filters = {
				removeKeyValueFromQuery: jest.fn().mockImplementation( () => ( {} ) ),
				submitRequest: jest.fn(),
			};
		} );

		afterEach( () => {
			global.window = windowHold;
		} );

		test( 'Should remove filter', () => {
			// Setup test.
			const $removeButton = $( '<button></button>' );
			const $pill = $( '<div data-filter-name="tribe_eventcategory[]"></div>' );
			$removeButton.closest = jest.fn().mockImplementation( () => $pill );
			const event = {
				data: {
					target: $removeButton,
					container: $(),
				},
			};

			// Test.
			tribe.filterBar.filterRemove.handleRemoveClick( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 1 );
		} );

		test( 'Should return early if pill does not have data-filter-name attribute', () => {
			// Setup test.
			const $removeButton = $( '<button></button>' );
			const $pill = $( '<div></div>' );
			$removeButton.closest = jest.fn().mockImplementation( () => $pill );
			const event = {
				data: {
					target: $removeButton,
					container: $(),
				},
			};

			// Test.
			tribe.filterBar.filterRemove.handleRemoveClick( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 0 );
		} );

		test( 'Should return early if data-filter-name attribute is empty', () => {
			// Setup test.
			const $removeButton = $( '<button></button>' );
			const $pill = $( '<div data-filter-name></div>' );
			$removeButton.closest = jest.fn().mockImplementation( () => $pill );
			const event = {
				data: {
					target: $removeButton,
					container: $(),
				},
			};

			// Test.
			tribe.filterBar.filterRemove.handleRemoveClick( event );

			// Confirm final states.
			expect( tribe.filterBar.filters.removeKeyValueFromQuery.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filters.submitRequest.mock.calls.length ).toBe( 0 );
		} );
	} );
} );
