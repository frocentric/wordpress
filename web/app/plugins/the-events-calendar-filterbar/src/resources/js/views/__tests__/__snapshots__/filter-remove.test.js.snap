// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Remove Selectors Should match snapshot 1`] = `"{\\"removeButton\\":\\"[data-js=\\\\\\"tribe-filter-bar-c-pill__remove-button\\\\\\"]\\",\\"pillFilterName\\":\\"[data-filter-name]\\"}"`;
